Migrate from KBuild to CMake

SebastianBoe opened the pull request:
RFC: [Replacing Make/Kbuild with CMake](https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-April/007538.html)
issue: https://github.com/zephyrproject-rtos/zephyr/issues/1488

This PR migrates Zephyr from Kbuild[1] to CMake[0]. This migration is an important step in a larger
effort to make Zephyr easy to use for application developers working on different platforms with different development environment needs.

Simplified, this change retains Kconfig as-is, and replaces all Makefiles with CMakeLists.txt. The DSL-like Make language that KBuild offers is replaced by a set of CMake extentions. These extentions have either provided simple one-to-one translations of KBuild features or introduced new concepts that replace KBuild concepts.

The PR will be merged when all blocker TODOs are complete or the TSC deems it ready.

Review is welcome and may or may not introduce new blocker or non-blocker TODOs.

NB: Github's "Commits" and "Files changed" feature is broken for some reason.
You'll need to use your local git client to review the changes.

# Blocker TODO

- [x] Port DTC overlay
- [x] Verify that all samples are similar enough before and after CMake
- [x] Verify on HW for a small selection of samples
- [x] Update win docs
- [x] Test building with Windows
- [x] Write migration text/announcement for mailinglist
- [x] Remove KBuild makefiles
- [x] Squash history of review fixes
- [x] Port bt-servr
- [x] openocd is not discovered correctly

# Non-blocker TODO
- [ ] Ninja is broken
- [ ] Port or deprecate xcc (Requires RHEL 4 or 5)
- [x] Update mac docs
- [ ] Update build instructions throughout docs using 'make` (~140 uses)
- [x] The CMake dependency is difficult to deal with

# References:

[0] https://cmake.org/overview/
[1] https://www.kernel.org/doc/Documentation/kbuild/makefiles.txt

Earlier PR https://github.com/zephyrproject-rtos/zephyr/pull/1107
Very early Gerrit preview: https://gerrit.zephyrproject.org/r/#/c/12672/

SebastianBoe committed:
Introduce cmake-based rewrite of KBuild

Introducing CMake is an important step in a larger effort to make
Zephyr easy to use for application developers working on different
platforms with different development environment needs.

Simplified, this change retains Kconfig as-is, and replaces all
Makefiles with CMakeLists.txt. The DSL-like Make language that KBuild
offers is replaced by a set of CMake extentions. These extentions have
either provided simple one-to-one translations of KBuild features or
introduced new concepts that replace KBuild concepts.

This is a breaking change for existing test infrastructure and build
scripts that are maintained out-of-tree. But for FW itself, no porting
should be necessary.

For users that just want to continue their work with minimal
disruption the following should suffice:

Install CMake 3.8.2+

Port any out-of-tree Makefiles to CMake.

Learn the absolute minimum about the new command line interface:

$ cd samples/hello_world
$ mkdir build && cd build
$ cmake -DBOARD=nrf52_pca10040 ..

$ cd build
$ make

PR: zephyrproject-rtos#4692
docs: http://docs.zephyrproject.org/getting_started/getting_started.html

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

Commit:
doc: application: fix stale sentence about Kbuild

The Overview section still mentions Kbuild, but the build system is
CMake now.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

nashif committed:
gitignore: add build directory to ignore list

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: support cmake in shippable

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
CODEOWNERS: add cmake owners

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: adapt for cmake

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sampels: remove obsoleted static_lib sample

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doxygen: ignore misc/util.h

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: update with CMake instructions

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
linker: use UTIL_LISTIFY macro for dealing with .a files

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kconfig: fix Qt header for building Qt based kconfig

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: use cmake to build object benchmarks

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: update shippable for new SDK

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

mbolivar commented:
Hi! I have a question about device tree overlay support in the CMake port.

In scripts/Makefile.lib, there is support for a pair of variables which may be set by the application (or, I suppose, on the command line), namely DTC_OVERLAY_FILE and DTC_OVERLAY_DIR. These allow for device tree overlays to be put in place on an application- or board-specific basis, respectively.

I'm looking for equivalent support in CMake, since a sample provided by mcuboot and other applications I work on rely on this functionality. However, I don't see support for DT overlays in dts.cmake or anywhere else.

Am I missing something?

SebastianBoe commented:
> Am I missing something?
You are not missing anything. I missed this feature.
Added to blocker list.

mbolivar reviewed (changes requested):
Hi, I'm having an issue flashing led_ws2812 on 96b_carbon:

```
 KBUILD_VERBOSE=1 ninja flash
[3/3] Flashing 96b_carbon
FAILED: zephyr/cmake/flash/CMakeFiles/flash 
cd /home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build && /usr/bin/cmake -E env DFUUTIL_PID=0483:df11 DFUUTIL_ALT=0 DFUUTIL_IMG=/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin DFUUTIL_DFUSE_ADDR= O=/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr KERNEL_ELF_NAME=zephyr.elf KERNEL_HEX_NAME=zephyr.hex KERNEL_BIN_NAME=zephyr.bin ARCH=arm BOARD_NAME=96b_carbon /home/mbolivar/.local/share/virtualenvs/zephyr-q52mD-rc/bin/python3 /home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py flash /home/mbolivar/zmp/zephyr/scripts/support/dfuutil.sh
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuERROR, status = 10
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
dfu-util: Only DfuSe file version 1.1a is supported
dfu-util: (for raw binary download, use the --dfuse-address option)
dfu-util -d,0483:df11 -l
dfu-util -d,0483:df11 -s :leave -a 0 -D /home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin
Traceback (most recent call last):
  File "/home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py", line 50, in <module>
    run(sys.argv[2], sys.argv[1], debug)
  File "/home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py", line 39, in run
    runner.run(command)
  File "/home/mbolivar/zmp/zephyr/scripts/support/runner/dfu.py", line 76, in run
    self.check_call(cmd)
  File "/home/mbolivar/zmp/zephyr/scripts/support/runner/core.py", line 257, in check_call
    subprocess.check_call(cmd)
  File "/usr/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['dfu-util', '-d,0483:df11', '-s', ':leave', '-a', '0', '-D', '/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin']' returned non-zero exit status 74.
ninja: build stopped: subcommand failed.
```

The root cause from the above trace is that DFUUTIL_DFUSE_ADDR is set to the empty string.

I checked boards/arm/96b_carbon/board.cmake, which contains the following:

```
set(FLASH_SCRIPT dfuutil.sh)

set_property(GLOBAL APPEND PROPERTY FLASH_SCRIPT_ENV_VARS
  DFUUTIL_PID=0483:df11
  DFUUTIL_ALT=0
  DFUUTIL_IMG=${PROJECT_BINARY_DIR}/${KERNEL_BIN_NAME}
  DFUUTIL_DFUSE_ADDR=${CONFIG_FLASH_BASE_ADDRESS}
  )
```

It would thus appear that CONFIG_ options are not available to CMake when board.cmake is included. That's strange, though, because kconfig.cmake is included earlier in the boilerplate file.

Any idea what's wrong?

mbolivar commented:
> You are not missing anything. I missed this feature.
> Added to blocker list.

Awesome, thank you.

SebastianBoe commented:
@mbolivar : Found the root cause. Will push a fix soon.

mbolivar reviewed (commented):
Sounds good; thanks for the fast response.

SebastianBoe committed:
dts: Break all dependencies of DTS on ext

If DTS does not depend on include paths determined in /ext then we can
move dts next to kconfig. Moving it next to kconfig means that DTS
CONFIG_ variables will be available at the same time as kconfig's
CONFIG_ variables which is the intended design and beneficial from a
usability perspective.

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

Commit:
doc: getting_started: add ref target for running hello world

This makes it possible to point users at a canonical location for how
to use zephyr-env.sh, etc.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: CMake updates for overview and creation docs

Update the initial application overview and the basic "how to create
an application" sections for the CMake transition. This is worth doing
on its own, and also enables other fixes and improvements to below
sections.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

SebastianBoe commented:
Fix pushed.

mbolivar commented:
Another question: I'm still building documentation locally with 'make htmldocs' from top level as described in doc/readme.rst. Is it planned to port this to CMake as well? I don't see this in either TODO list.

Commit:
doc: application: touch up cmake overview

Fix up some grammar issues and add more reST formatting.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

SebastianBoe commented:
> Another question: I'm still building documentation locally with 'make htmldocs' from top level as described in doc/readme.rst. Is it planned to port this to CMake as well? I don't see this in either TODO list.

Short answer; no.

The focus is on improving the Windows support and development environment support for application developers. There are no immediate plans to port Zephyr infrastructure that application developers are not expected to be using.

Commit:
doc: application: re-work CMakeLists.txt / configuration division

Fix a few issues related to the description of an application's
CMakeLists.txt file, and how that relates to its configuraiton.

Make sure the section "Application CMakeLists.txt" appears under the
parent "CMake" section, instead of on its own. The order in which
lines appear in the application CMakeLists.txt is important, and
that's not coming through clearly, so try to improve that. Document
how the values for BOARD and CONF_FILE are determined by
boilerplate.cmake here. Also document usage of KCONFIG_ROOT, as its
Kbuild-based equivalent is something that users ask about.

Merge some content from the following section "Application
Configuration" into the appropriate places, to keep the document flow
working. Add references in "Application Configuration" to definitions
provided previously in the document, for clarity.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
> Short answer; no.

That makes sense. So we'll just run `make` from the `doc` directory in the future, I'm guessing?

Commit:
doc: application: improve Kconfig setting documentation

Now that the above sections more clearly define the behavior of
CONF_FILE, the section on writing a .conf can be cleaned up and made
into a simple primer on syntax.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

nashif commented:
> That makes sense. So we'll just run make from the doc directory in the future, I'm guessing?

yep, that is easy enough I guess.

Commit:
doc: application: improvements to "Overriding Default Configuration"

Make this a bit less Unix-centric, changing the CMake invocation lines
to only use documented parameters that continue to work across all
CMake versions (-B doesn't work everywhere, and -H means "help" in
recent versions of CMake).

Handle some 80-column cleanliness. Add documentation about the search
key, /. Be a bit more explicit about the steps.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
I couldn't find a way to simply dismiss my review, but I'm no longer requesting changes.

SebastianBoe commented:
@mbolivar : If you were able to do any HW testing then I would be interested to know the results.

mbolivar commented:
@SebastianBoe I just did a quick sanity check that I could build and flash led_ws2812 on 96b_carbon and that the console output matched what I expected. This was enough to confirm for me that the configurations were being merged correctly, that it could find a SPI device, and that the flashing script glue was OK.

I figure you and @nashif are testing Nordic and Intel parts, so I just wanted to give a quick look at STM32 until the DT overlay support is added. Unfortunately the applications I care most about rely on that feature to build, so I won't be able to do the "real" testing I'd like to until that's at feature parity in CMake.

Commit:
doc: application: trivial fixups to "Rebuild an Application"

Make clear that a terminal is being used, and use a Sphinx :ref:
instead of a reST link.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: move build/run/debug docs after overview

This document's flow could be improved a bit. The overview section is
followed by somewhat dense reference material, which is then followed
by step-by-step instructions central to the application workflow.

Fix this up by moving the details to the end, and adjusting the
transitions between the sections a bit.

The diff looks like a mess, but this commit is mostly just moving
things around. There are also various grammar fixes incorporated from
review.

Reviewed-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

squash! doc: application: move build/run/debug docs after overview

Commit:
doc: application: trivial whitespace fix

Fix the white space around "Naming Conventions", which is currently
appearing on its own line.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

nashif reviewed (changes requested):
blocking merge until we are ready...

Commit:
doc: application: fix up "running an application"

Re-work this section for clarity, making it clearer where the division
between emulated and real hardware is, and cleaning up the
instructions. Also re-work for correctness, updating Kbuild-style
flash instructions to use CMake.

Also make sure users know they can flash and run from anywhere on the
system using cmake --build.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar reviewed (changes requested):
I can't build samples/hello_world on qemu_x86. The error when I run ninja is:

```
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: cannot find libzephyr.a
```

A full log is at https://hastebin.com/oheyexokuy.sql.

I can build for qemu_cortex_m3, however.

Commit:
doc: add zephyr-app-commands directive

Add extensions/zephyr to the documentation. This is where Sphinx
extensions customized for Zephyr will live.

Within, add application.py. This provides a directive,
zephyr-app-commands, which generates commands in the docs to build,
flash, debug, etc. an application. For now, these are Unix shell
specific. Later on, they can be customized to generate additional
formats, perhaps with extra options.

After this is used throughout the tree, doing this with an extension
enables global changes with changes to the directive implementation
only.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
boards: x86: convert make to cmake

Unify documentation formatting and use zephyr-app-commands where
applicable.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix up alternate CONF_FILE

Fix the instructions for when an alternate CONF_FILE is used.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: debugging fixes

Grammatical fixes caught in review of unrelated changes.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: define ELF

The ELF format and extension is an acronym that needs a
definition. Add it.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix QEMU capitalization

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix typo

s/remove/remote/

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: provide ddd install instructions

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: clarify a pronoun antecedent

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix "behavior" british spelling

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: revise app/zephyr link documentation

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: clarify KCONFIG_ROOT section

The documentation regarding application-specific Kconfig options is
unclear. Fix that.

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: add rest formatting to CMakeLists.txt

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fixup for app versus zephyr config

The line between Zephyr versus application is blurry since they share
a configuration, but try to disambiguate.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: merge configuration paragraphs

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: fix britishism regarding parentheses

Reported-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

SebastianBoe commented:
NB: Github's "Commits" and "Files changed" feature is broken for some reason.
You'll need to use your local git client to review the changes.

carlescufi committed:
doc: getting_started: Windows MSYS2 with CMake

Update the Windows MSYS2 instructions with the required CMake commands
used to build on this platform.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

jhedberg commented:
I took a peek at this and it looks mostly good, however it seems drivers/bluetooth is missing the CMake conversion. Is that known and still in the TODO?

nashif commented:
> I took a peek at this and it looks mostly good, however it seems drivers/bluetooth is missing the CMake conversion. Is that known and still in the TODO?

I think those have been completely missed! fixing now :)

dbkinder commented on doc/getting_started/installation_linux.rst:
```diff
@@ -52,23 +52,32 @@ Install the required packages in a Ubuntu host system with:
 
-   $ sudo apt-get install git make gcc g++ ncurses-dev gperf ccache\
-	 doxygen dfu-util device-tree-compiler python3-ply python3-pip
+   $ sudo apt-get install --no-install-recommends git ninja-build gperf \
```
Not all these changes for installed files are cmake related?

SebastianBoe replied:
Right.

The Ubuntu apt install command has changed slightly.

We use --no-install-recommends to have more control over exactly which packages we depend on.

This was tested using Docker, which uses a minimal Ubuntu install. This new apt-get install 
command is safer than the old one because it is explicitly listing some dependencies that
were assumed before.

carlescufi commented:
Tested latest branch on Windows/MSYS2 against latest docs. No issues.

SebastianBoe committed:
cmake: ci: Add more minions

remove this commit before merging with master

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
kbuild: Removed KBuild

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: Fixed subsys/net/lib/http

This fixes the CI failure of the sample lwm2m_client.

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

nashif commented:
@mbolivar does hello world build for you now?

nashif commented:
recheck

mbolivar commented:
> @mbolivar does hello world build for you now?

Yes.

mbolivar commented:
(no longer requesting changes; just dismissing my previous review)

dbkinder commented on doc/zephyr.doxyfile:
```diff
@@ -113,7 +113,9 @@ INPUT                  = \
 FILE_PATTERNS          = "*.c" "*.h" "*.S"
 RECURSIVE              = YES
-EXCLUDE                = ../include/spi_legacy.h ../include/net/http_legacy.h
+EXCLUDE                = ../include/spi_legacy.h
```
I think there's a missing ``\`` on this line, yes?

jhedberg commented:
Unless I'm missing something, it seems the default board selection on a per-app basis seems to get broken with this (i.e. I always have to specify BOARD=...), as does the configuration file selection based on the board (e.g. several Bluetooth samples had this). Is this still planned to be fixed?

jhedberg commented:
Correction, seems the configuration file selection based on the board seems to be working, however if I'm not mistaken BOARD=... now needs to be specified every time when you build?

nashif committed:
doc: fix doxygen EXCLUDE

Add backspace and fix spacing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> Correction, seems the configuration file selection based on the board seems to be working, however if I'm not mistaken BOARD=... now needs to be specified every time when you build?

yes, this was always an annoyance when a sample that is supposed to work with many boards has a default set to the authors favourite board. Most samples/tests were set to some qemu board, so you can replicate that by setting it in your environment which will give you the same results.

nashif commented:
> yes, this was always an annoyance when a sample that is supposed to work with many boards has a default set to the authors favourite board. Most samples/tests were set to some qemu board, so you can replicate that by setting it in your environment which will give you the same results.

btw, a sample that really only works on 1 board can still have a default, there is no issue with that, but most samples/tests we have run on multiple boards and architectures, so setting default was misleading

jhedberg commented:
I'm getting the following when building almost any Bluetooth sample app for qemu_x86. Our smallest app, the beacon gives:

`/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 20 bytes
`

whereas e.g. our mesh app gives:

`/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 8376 bytes
`

What's going on?

mbolivar commented:
> btw, a sample that really only works on 1 board can still have a default, there is no issue with that, but most samples/tests we have run on multiple boards and architectures, so setting default was misleading

FWIW (not 100% sure this is relevant), I've upstreamed how we address this in our own applications in the samples/drivers/led_ws2812 and led_lpd8806 samples, and this was propagated to cmake.

By merging in board-specific fragments into a high-level prj.conf (rather than prj.conf XOR prj_BOARD.conf), you can add the maintainer's favorite board on equal footing with other boards that people port things to.

Obviously if you can write a single prj.conf that always works, that's great, but if you can't then I think this is a reasonable compromise.

nashif commented:
> /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 8376 bytes
> 
> What's going on?

not sure, I do not get those at all...

mbolivar commented:
I'm seeing a new warning when building the docs (with `make -C doc htmldocs`):

```
../Kconfig.zephyr:11: warning: The symbol KERNELVERSION references the non-existent environment variable KERNELVERSION and will get the empty string as its value. If you're using Kconfiglib via 'make (i)scriptconfig', it should have set up the environment correctly for you. If you still got this message, that might be an error, and you should email ulfalizer a.t Google's email service.
```

Anyone else?

nashif commented:
> Anyone else?

that was always there...

mbolivar commented:
> that was always there...

Nope. Wasn't there when running `make htmldocs` from the top level makefile which is now gone, because it exported KERNELVERSION.

mbolivar commented:
> I'm getting the following when building almost any Bluetooth sample app for qemu_x86.

I'm seeing similar. E.g. samples/net/echo_server gives:

```
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: zephyr/zephyr_prebuilt.elf section `.noinit."/home/mbolivar/zmp/zephyr/subsys/net/ip/net_pkt.c".3' will not fit in region `MMU_LIST'
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region `MMU_LIST' overflowed by 8136 bytes
```

I tried checking out the latest commit in this branch that still has Kbuild and trying that, which is currently 35a411ebd, and I get a different error:

```
In file included from /home/mbolivar/zmp/zephyr/include/arch/x86/linker.ld:43:0,
                 from /home/mbolivar/zmp/zephyr/arch/x86/soc/ia32/linker.ld:46:
/home/mbolivar/zmp/zephyr/include/linker/linker-defs.h:108:2: error: #error "Expected NUM_KERNEL_OBJECT_FILES to be defined"
 #error "Expected NUM_KERNEL_OBJECT_FILES to be defined"
  ^~~~~
/home/mbolivar/zmp/zephyr/Makefile:868: recipe for target 'linker.cmd' failed
```

It's building fine on mainline.

@jhedberg, while we're here, I'm trying to update the documentation that references Make, and I ran into these lines in qemu_setup.rst:

```
   $ cd samples/net/echo_server
   $ make server
```

and

```
   $ cd samples/net/echo_client
   $ make client
```

Are the "server" and "client" targets still needed (and should they still work as targets to `ninja`) in the post-CMake world? If so I can update that once we sort out what's wrong with the build; if not, any recommendations on what to replace them with?

nashif commented:
> Nope. Wasn't there when running make htmldocs from the top level makefile which is now gone, because it exported KERNELVERSION.

yeah, the version handling and make it available to the docs is missing. Looking into that now.
I will also add a top-level Makefile to keep existing 'make htmldocs' working

nashif commented:
> I'm seeing similar. E.g. samples/net/echo_server gives

ok, this could be related to recent changes in master that were not correctly rebased I guess. @SebastianBoe  can you please take a look?

nashif committed:
cmake: fixed version handling

Use VERSION in top directory to specify version

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@SebastianBoe , the generated version.h does not match between master and the cmake PR:

in master:
```
#ifndef _KERNEL_VERSION_H_
#define _KERNEL_VERSION_H_

#define ZEPHYR_VERSION_CODE 67939
#define ZEPHYR_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define KERNELVERSION \
0x01096300
#define KERNEL_VERSION_NUMBER     0x010963
#define KERNEL_VERSION_MAJOR      1
#define KERNEL_VERSION_MINOR      9
#define KERNEL_PATCHLEVEL         99
#define KERNEL_VERSION_STRING     "1.9.99"

#endif /* _KERNEL_VERSION_H_ */
```

in cmake PR:

```
#ifndef _KERNEL_VERSION_H_
#define _KERNEL_VERSION_H_

#define ZEPHYR_VERSION_CODE 65793
#define ZEPHYR_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define KERNELVERSION          65793
#define KERNEL_VERSION_NUMBER  16843008
#define KERNEL_VERSION_MAJOR   1
#define KERNEL_VERSION_MINOR   9
#define KERNEL_PATCHLEVEL      99
#define KERNEL_VERSION_STRING  "1.9.99"

#endif /* _KERNEL_VERSION_H_ */
```

nashif committed:
doc: fixed version parsing for docs

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
build: add top-level Makefile for docs

for docs and everything that is not covered by cmake.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: call 'make htmldocs' from root

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
ok, version handling fixed I hope: #4778

mbolivar commented:
I'm also getting unexpected output when I `ninja run` hello_world for qemu_x86:

```
$ ninja run
[91/92] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: qemu32,+nx,+pae
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin
```

When building for qemu_cortex_m3, I see the expected output (Hello World! arm)

mbolivar commented:
> ok, version handling fixed I hope: #4778

It's working for me here; thanks. Left a couple comments on that PR.

mbolivar commented:
@SebastianBoe: flash support for nrf52840_pca10056 appears to have regressed, but I'd like to confirm.

A Makefile.board was previously available for that board, and I tested it with the existing flash scripts fairly recently. But there's no board.cmake in this branch to keep that support.

Since this is a Nordic board, I want to ask if this was intentional.

jhedberg reviewed (changes requested):
The original samples/bluetooth/mesh_demo had the following in its Makefile:

ifneq ($(NODE_ADDR),)
CFLAGS += -DNODE_ADDR=$(NODE_ADDR)
endif

This is an essential part of the app since it doesn't rely on external provisioning, i.e. every image needs a unique address. Basically you would always build it using "make NODE_ADDR=0x0001" or similar. Unless I'm missing something, I don't see this taken into account in the cmake migration (I'd propose a patch but my cmake knowledge is still way too limited).

jukkar commented on samples/net/rpl-node/Makefile:
```diff
@@ -1,15 +0,0 @@
-# Makefile - RPL node test application
```
There is no CMakeLists.txt file for rpl-node sample.

SebastianBoe replied:
Are you familiar with this sample, or did you just happen to notice it?

Any assistance here would be appreciated.
https://github.com/zephyrproject-rtos/zephyr/issues/4791

SebastianBoe replied:
Ported.

Now it is broken in both KBuild and CMake, whereas before it was absent in CMake and broken in KBuild.

https://github.com/zephyrproject-rtos/zephyr/pull/4792

jhedberg commented:
The MMU_LIST overflow doesn't seem to happen with the default make generator, but does happen when I use -GNinja.

Another observation: I didn't at first realize that BOARD must be part of the environment during both the cmake and the make step. I was initially only giving it to cmake, and then running make without the environment variable would cause the build to fail with slightly confusing errors.

jukkar commented:
> Are the "server" and "client" targets still needed (and should they still work as targets to ninja) in the post-CMake world? If so I can update that once it's sorted out what's wrong with the build; if not, any recommendations on what to replace them with?

I am hoping I would be able to run two qemu's against each other after this migration. But it seems that samples/net/common/Makefile.ipstack is been removed and there is not substitute using cmake. This is major issue for testing. Also this prevents simulated 802.15.4 environment using two qemu's from working.

jhedberg commented:
Question: how do I run qemu with the make generator? I can successfully build an image with that generator using simply "make", but "make run" doesn't seem to work. The new docs say that "ninja run" should work, but as I've previously mentioned ninja doesn't work for me on any Bluetooth sample because of the MMU_LIST overflow issue.

jukkar commented:
I am using Fedora 26 and I have cmake 3.9.1 installed. Could the cmake be part of SDK and zephyr to use that directly so that user would not need to downgrade/fiddle with cmake sources?

jukkar commented:
> But it seems that samples/net/common/Makefile.ipstack is been removed and there is not substitute using cmake. This is major issue for testing. Also this prevents simulated 802.15.4 environment using two qemu's from working.

After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.

SebastianBoe commented:
> Question: how do I run qemu with the make generator? I can successfully build an image with that generator using simply "make", but "make run" doesn't seem to work. The new docs say that "ninja run" should work, but as I've previously mentioned ninja doesn't work for me on any Bluetooth sample because of the MMU_LIST overflow issue. -- @jhedberg 

You are doing it correctly. "make run" should work. If you have found a corner where it doesn't please provide steps to reproduce. Run "make usage" to see all available command line targets.

SebastianBoe committed:
cmake: Ported samples/net/rpl-node

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

jukkar commented on doc/getting_started/installation_linux.rst:
```diff
@@ -52,23 +52,32 @@ Install the required packages in a Ubuntu host system with:
+   $ mkdir $HOME/work_dir && cd $HOME/work_dir
+   $ wget https://cmake.org/files/v3.8/cmake-3.8.2-Linux-x86_64.sh
+   $ yes | sh cmake-3.8.2-Linux-x86_64.sh | cat
+   $ echo "export PATH=$PWD/cmake-3.8.2-Linux-x86_64/bin:\$PATH" >> $HOME/.bashrc
```
Instead of putting the PATH to .bashrc, better would be to direct to `$HOME/.zephyrrc` as then the cmake would not conflict with system installed cmake if that is some newer version etc. The .zephyrrc is sourced by zephyr-env.sh.

erwango commented:
Seems that we lost the use of SDK openocd.
When running "make flash", it uses my system openocd (older than the SDK one). Then for boards having flash dependencies to SDK version, flash doesn't work.

jukkar commented:
> After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.

Verifying that "make server" and "make client" works for net echo-server and echo-client samples. I was not able to use ninja as I get the MMU_LIST overflow issue.

erwango commented:
Another point:
This is surely a commonly admitted cmake way of working and I just need to get used to it, but it feels a bit painful to type 4 commands instead of one simple "make BOARD=<board>".
Is that foreseen to have an alias to at least avoid the "mkdir build/cd build"?

ydamigos commented:
Running cmake fails to create the needed files (make or ninja) to build sample subsys/usb/cdc_acm for 96b_carbon because the options  USB_DEVICE_VID and USB_DEVICE_PID don't have a default value. Running menuconfig is not possible. Is it possible to build the sample without changing the prj.conf file?

erwango reviewed (changes requested):
Seems that we lost the use of SDK openocd.
When running "make flash", it uses my system openocd (older than the SDK one). Then for boards having flash dependencies to SDK version, flash doesn't work.

Last, booting banner disappeared. Not a blocker, but it is a nice feature to see flashed binary is the epxected one (last build):
"***** BOOTING ZEPHYR OS v1.9.99 - BUILD: Nov  8 2017 11:07:46 *****"

SebastianBoe commented:
>Another point:
This is surely a commonly admitted cmake way of working and I just need to get used to it, but it feels a bit painful to type 4 commands instead of one simple "make BOARD=".
Is that foreseen to have an alias to at least avoid the "mkdir build/cd build"?

It is painful indeed. I use this alias personally for clean builds:

alias cb='rm -rf b && cmake -Bb -H. && cmake --build b'

SebastianBoe committed:
cmake: Introduce a default VID and PID

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
> Running cmake fails to create the needed files (make or ninja) to build sample subsys/usb/cdc_acm for 96b_carbon because the options USB_DEVICE_VID and USB_DEVICE_PID don't have a default value. Running menuconfig is not possible. Is it possible to build the sample without changing the prj.conf file?

Fixed in https://github.com/zephyrproject-rtos/zephyr/pull/4795

carlescufi committed:
doc: cmake: Update macOS build instruction

Update macOS build instructions so that they reflect the transition to
CMake.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

erwango commented:
>> It is painful indeed. I use this alias personally for clean builds:

> alias cb='rm -rf b && cmake -Bb -H. && cmake --build b'

Thanks for the hint! Wouldn't it be valuable to provide a full set of official zephyr aliases to avoid issues like build issues do to lack of proper build clean and alike?

erwango commented on ext/hal/st/stm32cube/CMakeLists.txt:
```diff
@@ -0,0 +1,120 @@
+# STM32Cube HAL and LL APIs
+zephyr_compile_definitions(
+    -D${STM32CUBE_CPU}
+	-DUSE_HAL_DRIVER
```
nitpick: alignment

SebastianBoe committed:
cmake: Fixed NODE_ADDR

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
> This is an essential part of the app since it doesn't rely on external provisioning, i.e. every image needs a unique address. Basically you would always build it using "make NODE_ADDR=0x0001" or similar. Unless I'm missing something, I don't see this taken into account in the cmake migration (I'd propose a patch but my cmake knowledge is still way too limited).
@jhedberg 

Fixed.
https://github.com/zephyrproject-rtos/zephyr/pull/4797

SebastianBoe commented:
> Thanks for the hint! Wouldn't it be valuable to provide a full set of official zephyr aliases to avoid issues like build issues do to lack of proper build clean and alike?

Perhaps we could do so in the future. For now users will have to figure this out on their own.

SebastianBoe committed:
cmake: ext/hal/st/stm32cube: Fixed formatting and removed hal_uart

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe commented on ext/hal/st/stm32cube/CMakeLists.txt:
```diff
@@ -0,0 +1,120 @@
+zephyr_sources(stm32l4xx/drivers/src/stm32l4xx_hal_rcc.c)
+zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_STM32_CUBE stm32l4xx/drivers/src/stm32l4xx_ll_utils.c)
+zephyr_sources_ifdef(CONFIG_PWM stm32l4xx/drivers/src/stm32l4xx_hal_tim.c)
+zephyr_sources_ifdef(CONFIG_SERIAL_HAS_DRIVER stm32l4xx/drivers/src/stm32l4xx_hal_uart.c)
```
Fixed all comments in this file:
https://github.com/zephyrproject-rtos/zephyr/pull/4798

SebastianBoe committed:
cmake: Use Make for examples instead of Ninja

Ninja is not tested as well as Make yet, so for now, recommend using
Make.

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

carlescufi committed:
doc: cmake: Update the Linux CMake instructions

We no longer want to pin users to 3.8.2. Instead we tolerate the warning
and therefore ask users to get the latest CMake package from their
distro.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@erwango aliases would be great. Would you give it a go in a PR to this branch?

galak reviewed (changes requested):
In the 'Introduce cmake-based rewrite of KBuild' would be nice to have simple usage details and point to the full docs in the commit message.

SebastianBoe committed:
cmake: Use -serial bt-server when CONFIG_BT

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

erwango commented:
@carlescufi: I'll have a try

mbolivar commented:
>> After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.
> Verifying that "make server" and "make client" works for net echo-server and echo-client samples. I was not able to use ninja as I get the MMU_LIST overflow issue.

@jukkar -- OK, so "make server" and "make client" works for Makefile generated systems but not Ninja ones. I've heard on IRC that we're going to start with Make and switch to Ninja when that's fixed.

We should figure out what to document at first, then. I'm working on porting all the board docs that reference "make flash" etc to the new system (https://github.com/zephyrproject-rtos/zephyr/pull/4805). I can start this with make and switch to Ninja in a follow on after it's working, or I can stick with Ninja. This also affects the other PRs that are updating other documents.

I'd like to get a verdict on what to do -- @nashif, @SebastianBoe, @carlescufi could you please let me know to start with Make or Ninja?

carlescufi committed:
doc: cmake: Unify getting started guides

Unify the format and mechanisms used in the different Getting Started
guides so that they are consistent in:

- The way Kconfig is built
- Avoiding using -B and -H CMake options
- -DBOARD instead of export

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

pfalcon commented:
> could you please let me know to start with Make or Ninja?

My 2 cents: people used make before, of course people will use make after. So, it's kinda strange that ninja is even being mentioned when the current goal is to get CMake into the 1.10 release. Please make make work, anything else should wait, thanks ;-).

SebastianBoe commented:
@mbolivar start with Make.

nashif commented:
> I'd like to get a verdict on what to do -- @nashif, @SebastianBoe, @carlescufi could you please let me know to start with Make or Ninja?

make

SebastianBoe committed:
cmake: Pass GDB env var to flasher script

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

carlescufi commented:
@mbolivar make, but do not remove the ninja code from the Sphinx script or the doc, simply default to make for now.

mbolivar commented:
ACK

carlescufi commented:
@galak yep, I agree with that.

SebastianBoe committed:
cmake: Fixed samples/net/rpl-node

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

mbolivar commented:
> @SebastianBoe: flash support for nrf52840_pca10056 appears to have regressed, but I'd like to confirm.

Any comment on this one?

SebastianBoe committed:
cmake: Fixed openocd

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: Fixed flashing of nrf52840_pca10056

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
> Any comment on this one? --  @mbolivar 

Fixed. https://github.com/zephyrproject-rtos/zephyr/pull/4811

SebastianBoe commented:
> In the 'Introduce cmake-based rewrite of KBuild' would be nice to have simple usage details and point to the full docs in the commit message. -- @galak 

Is this msg ok?
https://github.com/SebastianBoe/zephyr/commit/f5b0face7a9cf4d0ee1b3b717ba9992ae37ec9c3

SebastianBoe commented:
@erwango : Please click the approve button if you approve of the requested changes.

carlescufi commented:
Commit message is fine, but please use the conventions in the getting started files. Single line mkdir && cd and -DBOARD. Also link to getting started on docs.zephyrproject.org

SebastianBoe commented:
> Commit message is fine, but please use the conventions in the getting started files. Single line mkdir && cd and -DBOARD. Also link to getting started on docs.zephyrproject.org

Done (on it's way).

galak commented:
Can we hide zephyr_library() inside something like zephyr_library_sources()?

Looking at an example in boards/arm/96b_carbon/CMakeLists.txt

galak commented:
Why do we have some boards with zephyr_library() and some just have zephyr_sources()?

boards/arm/cc3220sf_launchxl/CMakeLists.txt vs boards/arm/96b_carbon/CMakeLists.txt

mbolivar commented:
> Any comment on this one? -- @mbolivar
> Fixed. #4811

Thanks @SebastianBoe!

nashif committed:
cmake: use zephyr_library_* for all boards

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> Why do we have some boards with zephyr_library() and some just have zephyr_sources()?
> 
> boards/arm/cc3220sf_launchxl/CMakeLists.txt vs boards/arm/96b_carbon/CMakeLists.txt

Fixed in #4820

andrewboie commented on CONTRIBUTING.rst:
```diff
@@ -233,18 +233,15 @@ run the same tests the CI system runs, follow these steps from within your
-    $ export PREBUILT_HOST_TOOLS=${ZEPHYR_BASE}/bin
-    $ export USE_CCACHE=1
-    $ ./scripts/sanitycheck
+    $ ./scripts/sanitycheck --ccache
```
i really hope we didn't purge the USE_CCACHE variable...
It might be better to direct users to that instead of having to pass "--ccache" every time

SebastianBoe replied:
We did. I think I know what kind of behaviour you want. I'll push a fix.

SebastianBoe replied:
Fixed: https://github.com/zephyrproject-rtos/zephyr/pull/4837
